include/sys/scheduler.h

```diff
@@ -0,0 +1,171 @@
+ * - if the callback function is provided, run the callback function with
+ *   the provided object and context pointers
+ * - Set the thread's swap return values to swap_retval and swap_data
+ * - un-pend and ready the thread, but do not invoke the scheduler.
+ *
+ * It is up to the caller to implement locking such that the return value of
+ * this function (whether a thread was woken up or not) does not immediately
+ * become stale.
+ *
+ * Repeated calls to this function until it returns false is a suitable
+ * way to wake all threads on the queue.
+ *
+ * It is up to the caller to implement locking such that the return value of
+ * this function does not immediately become stale. Calls to wait and wake on
+ * the same wait_q object must have synchronization. Calling this without
+ * holding any spinlock is a sign that this API is not being used properly.
+ *
+ * @param wait_q Wait queue to wake up the highest prio thread
+ * @param swap_retval Swap return value for woken thread
+ * @param swap_data Data return value to supplement swap_retval. May be NULL.
+ * @param cb Callback function, or NULL in which case no callback is invoked.
+ * @param obj Kernel object associated with this wake operation, passed to cb.
+ *        Only used by the callback. May be NULL.
+ * @param context Additional context pointer associated with this wake
+ *        operation. Only used by the callback. May be NULL.
+ * @retval true If a thread waa woken up
+ * @retval false If the wait_q was empty
+ */
+bool k_sched_callback_wake(_wait_q_t *wait_q, int swap_retval, void *swap_data,
+			   k_sched_wake_cb_t cb, void *obj, void *context);
+
+/**
+ * Wake up a thread pending on the provided wait queue
+ *
+ * Convenient function to k_sched_callback_wake(), for the very common case
+ * where no callback is required.
+ *
+ * @param wait_q Wait queue to wake up the highest prio thread
+ * @param swap_retval Swap return value for woken thread
+ * @param swap_data Data return value to supplement swap_retval. May be NULL.
+ * @retval true If a thread waa woken up
+ * @retval false If the wait_q was empty
+ */
+static inline bool k_sched_wake(_wait_q_t *wait_q, int swap_retval,
+				void *swap_data)
+{
+	return k_sched_callback_wake(wait_q, swap_retval, swap_data,
+				     NULL, NULL, NULL);
+}
+
+/**
+ * Wake up all threads pending on the provided wait queue
+ *
+ * Convenience function to invoke k_sched_wake() on all threads in the queue
+ * until there are no more to wake up.
+ *
+ * @param wait_q Wait queue to wake up the highest prio thread
+ * @param swap_retval Swap return value for woken thread
+ * @param swap_data Data return value to supplement swap_retval. May be NULL.
+ * @retval true If any threads were woken up
+ * @retval false If the wait_q was empty
+ */
+static inline bool k_sched_wake_all(_wait_q_t *wait_q, int swap_retval,
+				    void *swap_data)
+{
+	bool woken = false;
+
+	while (k_sched_wake(wait_q, swap_retval, swap_data)) {
+		woken = true;
+	}
```

Comment:
This (releasing the lock between each thread) is the way z_unpend_all() works currently, but this probably deserves some thought as to whether we want this to be an atomic operation (hold the lock while you release all threads simultaneously).  The latter is something that can't be emulated by the caller, where they could write the loop themselves.

The question is do we want that?  I dunno, but it would be more cleanly specified, less surprising, and easy to do right here.

Reply:
I had to do something like that in #29618 where I had a list of waiting threads and wanted to wake a subset of them, done by iterating over the set and extracting the ones I needed to a separate list under lock before releasing each thread (because the per-thread operation was a reschedule point; see #29610).  A waitq has different semantics, so maybe it isn't appropriate for this interface.

Reply:
the idea with these APIs is that there are two locks involved:
- `sched_spinlock`, which is held for individual calls to k_sched_wake() which protects the integrity of the scheduling data structures and serializes the state of individual threads that were woken up. This keeps the scheduler consistent. This is all internal to these APIs.
- Another, IPC-specific lock which establishes atomicity for whatever IPC operations being done. It's expected that this second lock is held by the caller across all `k_sched_wake()` and `k_sched_wake_all()` calls, until it's time to call `k_sched_invoke()` with that lock.

I put in the documentation for `k_sched_wake()`:

```
 * It is up to the caller to implement locking such that the return value of
 * this function does not immediately become stale. Calls to wait and wake on
 * the same wait_q object must have synchronization. Calling this without
 * holding any spinlock is a sign that this API is not being used properly.
```

@andyross so in other words, we actually do have:

> hold the lock while you release all threads simultaneously

As the second lock will cover this case with respect to users of whatever IPC object is being built.
I don't think we need to hold `sched_spinlock()` across these calls, as the second lock should be enough.

Let me know if this approach is workable.